File: Chroma-Core/src/main/java/buttondevteam/lib/architecture/ButtonPlugin.java
package buttondevteam.lib.architecture;

import buttondevteam.buttonproc.HasConfig;
import buttondevteam.core.ComponentManager;
import buttondevteam.lib.TBMCCoreAPI;
import buttondevteam.lib.chat.Command2MC;
import buttondevteam.lib.chat.ICommand2MC;
import lombok.AccessLevel;
import lombok.Getter;
import org.bukkit.configuration.InvalidConfigurationException;
import org.bukkit.configuration.file.FileConfiguration;
import org.bukkit.configuration.file.YamlConfiguration;
import org.bukkit.plugin.java.JavaPlugin;

import java.io.File;
import java.io.IOException;
import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;
import java.util.Arrays;
import java.util.Optional;
import java.util.Stack;

@HasConfig(global = true)
public abstract class ButtonPlugin extends JavaPlugin {
	@Getter //Needs to be static as we don't know the plugin when a command is handled
	private static Command2MC command2MC = new Command2MC();
	@Getter(AccessLevel.PROTECTED)
	private IHaveConfig iConfig;
	private CommentedConfiguration yaml;
	@Getter(AccessLevel.PROTECTED)
	private IHaveConfig data; //TODO
	/**
	 * Used to unregister components in the right order - and to reload configs
	 */
	@Getter
	private Stack<Component<?>> componentStack = new Stack<>();
	;

	protected abstract void pluginEnable();

	/**
	 * Called after the components are unregistered
	 */
	protected abstract void pluginDisable();

	/**
	 * Called before the components are unregistered
	 */
	protected void pluginPreDisable() {
	}

	@Override
	public final void onEnable() {
		if (!loadConfig()) {
			getLogger().warning("Please fix the issues and restart the server to load the plugin.");
			return;
		}
		try {
			pluginEnable();
		} catch (Exception e) {
			TBMCCoreAPI.SendException("Error while enabling plugin " + getName() + "!", e);
		}
		if (configGenAllowed(this)) //If it's not disabled (by default it's not)
			IHaveConfig.pregenConfig(this, null);
	}

	private boolean loadConfig() {
		var config = getConfig();
		if (config == null)
			return false;
		var section = config.getConfigurationSection("global");
		if (section == null) section = config.createSection("global");
		if (iConfig != null) iConfig.reset(section);
		else iConfig = new IHaveConfig(section, this::saveConfig);
		return true;
	}

	@Override
	public final void onDisable() {
		try {
			pluginPreDisable();
			ComponentManager.unregComponents(this);
			pluginDisable();
			if (ConfigData.saveNow(getConfig()))
				getLogger().info("Saved configuration changes.");
			iConfig = null; //Clearing the hashmap is not enough, we need to update the section as well
			getCommand2MC().unregisterCommands(this);
		} catch (Exception e) {
			TBMCCoreAPI.SendException("Error while disabling plugin " + getName() + "!", e);
		}
	}

	@Override
	public void reloadConfig() {
		tryReloadConfig();
	}

	public boolean tryReloadConfig() {
		if (!justReload()) return false;
		loadConfig();
		componentStack.forEach(c -> Component.updateConfig(this, c));
		return true;
	}

	public boolean justReload() {
		if (yaml != null && ConfigData.saveNow(getConfig())) {
			getLogger().warning("Saved pending configuration changes to the file, didn't reload. Apply your changes again.");
			return false;
		}
		var file = new File(getDataFolder(), "config.yml");
		var yaml = new CommentedConfiguration(file);
		if (file.exists()) {
			try {
				yaml.load(file);
			} catch (IOException | InvalidConfigurationException e) {
				getLogger().warning("Failed to load config! Check for syntax errors.");
				e.printStackTrace();
				return false;
			}
		}
		this.yaml = yaml;
		var res = getTextResource("configHelp.yml");
		if (res == null)
			return true;
		var yc = YamlConfiguration.loadConfiguration(res);
		for (var kv : yc.getValues(true).entrySet())
			if (kv.getValue() instanceof String)
				yaml.addComment(kv.getKey(), Arrays.stream(((String) kv.getValue()).split("\n"))
					.map(str -> "# " + str.trim()).toArray(String[]::new));
		return true;
	}

	@Override
	public FileConfiguration getConfig() {
		if (yaml == null)
			justReload(); //TODO: If it fails to load, it'll probably throw an NPE
		return yaml;
	}

	@Override
	public void saveConfig() {
		try {
			if (yaml != null)
				yaml.save();
		} catch (Exception e) {
			TBMCCoreAPI.SendException("Failed to save config", e);
		}
	}

	/**
	 * Registers command and sets its plugin.
	 *
	 * @param command The command to register
	 */
	protected void registerCommand(ICommand2MC command) {
		command.registerToPlugin(this);
		getCommand2MC().registerCommand(command);
	}

	@Retention(RetentionPolicy.RUNTIME)
	@Target(ElementType.TYPE)
	public @interface ConfigOpts {
		boolean disableConfigGen() default false;
	}

	public static boolean configGenAllowed(Object obj) {
		return !Optional.ofNullable(obj.getClass().getAnnotation(ConfigOpts.class))
			.map(ConfigOpts::disableConfigGen).orElse(false);
	}
}

File: Chroma-Core/src/main/java/buttondevteam/lib/architecture/ButtonPlugin.kt
package buttondevteam.lib.architecture

import buttondevteam.buttonproc.HasConfig
import buttondevteam.core.ComponentManager
import buttondevteam.lib.TBMCCoreAPI
import buttondevteam.lib.chat.Command2MC
import buttondevteam.lib.chat.ICommand2MC
import org.bukkit.configuration.InvalidConfigurationException
import org.bukkit.configuration.file.FileConfiguration
import org.bukkit.configuration.file.YamlConfiguration
import org.bukkit.plugin.PluginDescriptionFile
import org.bukkit.plugin.java.JavaPlugin
import org.bukkit.plugin.java.JavaPluginLoader
import java.io.File
import java.io.IOException
import java.util.*
import java.util.function.Consumer

@HasConfig(global = true)
abstract class ButtonPlugin : JavaPlugin {
    protected val iConfig = IHaveConfig(::saveConfig, section)
    private var yaml: YamlConfiguration? = null

    /**
     * May change if the config is reloaded
     */
    private val section get() = config.getConfigurationSection("global") ?: config.createSection("global")

    protected val data //TODO
        : IHaveConfig? = null

    /**
     * Used to unregister components in the right order - and to reload configs
     */
    val componentStack = Stack<Component<*>>()

    // Support testing with the protected constructor (MockBukkit)
    constructor() : super()
    protected constructor(loader: JavaPluginLoader, description: PluginDescriptionFile, dataFolder: File, file: File) : super(loader, description, dataFolder, file)

    protected abstract fun pluginEnable()

    /**
     * Called after the components are unregistered
     */
    protected abstract fun pluginDisable()

    /**
     * Called before the components are unregistered
     */
    protected open fun pluginPreDisable() {}
    override fun onEnable() {
        if (!tryReloadConfig()) {
            logger.warning("Please fix the issues and restart the server to load the plugin.")
            return
        }
        try {
            pluginEnable()
        } catch (e: Exception) {
            TBMCCoreAPI.SendException("Error while enabling plugin $name!", e, this)
        }
        if (configGenAllowed(this)) //If it's not disabled (by default it's not)
            IHaveConfig.pregenConfig(this, null)
    }

    override fun onDisable() {
        try {
            pluginPreDisable()
            ComponentManager.unregComponents(this)
            pluginDisable()
            if (ConfigData.saveNow(config)) logger.info("Saved configuration changes.")
            command2MC.unregisterCommands(this)
        } catch (e: Exception) {
            TBMCCoreAPI.SendException("Error while disabling plugin $name!", e, this)
        }
    }

    override fun reloadConfig() {
        tryReloadConfig()
    }

    fun tryReloadConfig(): Boolean {
        if (!saveAndReloadYaml()) return false
        iConfig.reload(section)
        componentStack.forEach(Consumer { c -> c.updateConfig() })
        return true
    }

    /**
     * Returns whether the config was loaded successfully. Otherwise, an empty config is used.
     */
    private fun saveAndReloadYaml(): Boolean {
        if (isConfigLoaded && ConfigData.saveNow(config)) {
            logger.warning("Saved pending configuration changes to the file, didn't reload. Apply your changes again.")
            return false
        }
        val file = File(dataFolder, "config.yml")
        val yaml = YamlConfiguration()
        if (file.exists()) {
            try {
                yaml.load(file)
            } catch (e: IOException) {
                logger.warning("Failed to load config! Check for syntax errors.")
                e.printStackTrace()
                return false
            } catch (e: InvalidConfigurationException) {
                logger.warning("Failed to load config! Check for syntax errors.")
                e.printStackTrace()
                return false
            }
        }
        this.yaml = yaml
        val res = getTextResource("configHelp.yml") ?: return true
        val yc = YamlConfiguration.loadConfiguration(res)
        for ((key, value) in yc.getValues(true))
            if (value is String) yaml.setComments(
                key.replace(".generalDescriptionInsteadOfAConfig", ""),
                value.split("\n").map { str -> "# " + str.trim { it <= ' ' } }
            )
        return true
    }

    override fun getConfig(): FileConfiguration {
        if (!isConfigLoaded) saveAndReloadYaml()
        return yaml ?: YamlConfiguration() //Return a temporary instance
    }

    override fun saveConfig() {
        if (isConfigLoaded) super.saveConfig()
    }

    val isConfigLoaded get() = yaml != null

    /**
     * Registers command and sets its plugin.
     *
     * @param command The command to register
     */
    fun registerCommand(command: ICommand2MC) {
        command.registerToPlugin(this)
        command2MC.registerCommand(command)
    }

    @Retention(AnnotationRetention.RUNTIME)
    @Target(AnnotationTarget.ANNOTATION_CLASS, AnnotationTarget.CLASS)
    annotation class ConfigOpts(val disableConfigGen: Boolean = false)
    companion object {
        //Needs to be static as we don't know the plugin when a command is handled
        @JvmStatic
        val command2MC = Command2MC()
        fun configGenAllowed(obj: Any): Boolean {
            return !Optional.ofNullable(obj.javaClass.getAnnotation(ConfigOpts::class.java))
                .map { it.disableConfigGen }.orElse(false)
        }
    }
}